tests/particles/test_decorators.py

```diff
@@ -748,3 +748,31 @@ def function_with_pos_and_var_positional_arguments(
     expected = (a, args, Particle(particle))
     actual = function_with_pos_and_var_positional_arguments(a, *args, particle=particle)
     assert actual == expected
+
+
+@pytest.mark.parametrize(
+    ("criteria", "kwargs", "exception"),
+    [
+        ({"require": "ion"}, {"particle": ["p+", "He"]}, ParticleError),
+        ({"require": "isotope"}, {"particle": ["p+", "He"]}, ParticleError),
+        (
+            {"any_of": {"lepton", "neutrino"}},
+            {"particle": ["p+", "alpha"]},
+            ParticleError,
+        ),
+        ({"exclude": "lepton"}, {"particle": ["p+", "e-"]}, ParticleError),
+    ],
+)
+def test_particle_categorization_of_particle_lists(
+    criteria: dict[str, str | Iterable[str]],
+    kwargs: dict[str, ParticleListLike],
+    exception: Exception,
+) -> None:
+    @particle_input(**criteria)  # type: ignore[arg-type]
+    def get_particle(
+        particle: ParticleLike,
+    ) -> Particle | ParticleList | CustomParticle:
+        return particle  # type: ignore[return-value]
```

Comment:
`@particle_input` dynamically transforms objects passed as arguments to functions, it causes problems for _static_ type checkers.  Hence, the `# type: ignore` comments.